Core/src/GapicRequestWrapper.php

```diff
@@ -0,0 +1,189 @@
+use Google\Protobuf\Internal\Message;
+use Google\Rpc\BadRequest;
+use Google\Rpc\Code;
+use Google\Rpc\RetryInfo;
+
+/**
+ * @internal
+ * The GapicRequestWrapper is responsible for delivering REST/gRPC requests.
+ */
+class GapicRequestWrapper
+{
+    /**
+     * @var Serializer A serializer used to encode responses.
+     */
+    private $serializer;
+
+    /**
+     * @var array Map of error metadata types to RPC wrappers.
+     */
+    private $metadataTypes = [
+        'google.rpc.retryinfo-bin' => RetryInfo::class,
+        'google.rpc.badrequest-bin' => BadRequest::class
+    ];
+
+    /**
+     * @param array $config [optional] {
+     *     @type Serializer $serializer A serializer used to encode responses.
+     * }
+     */
+    public function __construct(array $config = [])
+    {
+        $this->serializer = $config['serializer'] ?? new Serializer;
+    }
+
+    /**
+     * Deliver the request.
+     *
+     * @param callable $request The request to execute.
+     * @param array $args The arguments for the request.
+     * @return array
+     * @throws Exception\ServiceException
+     */
+    public function send(callable $request, array $args)
+    {
+        try {
+            $response = call_user_func_array($request, $args);
+            return $this->handleResponse($response);
+        } catch (\Exception $ex) {
+            if ($ex instanceof ApiException) {
+                throw $this->convertToGoogleException($ex);
+            }
+
+            throw $ex;
+        }
+    }
+
+    /**
+     * Serializes a gRPC response.
+     *
+     * @param mixed $response
+     * @return \Generator|OperationResponse|array|null
+     */
+    private function handleResponse($response)
+    {
+        if ($response instanceof PagedListResponse) {
+            $response = $response->getPage()->getResponseObject();
+        }
+
+        if ($response instanceof Message) {
+            return $this->serializer->encodeMessage($response);
+        }
+
+        if ($response instanceof OperationResponse) {
+            return $response;
+        }
+
+        if ($response instanceof ServerStream) {
+            return $this->handleStream($response);
+        }
+
+        return null;
+    }
+
+    /**
+     * Handles a streaming response.
+     *
+     * @param ServerStream $response
+     * @return \Generator|array|null
+     * @throws Exception\ServiceException
+     */
+    private function handleStream($response)
+    {
+        try {
+            foreach ($response->readAll() as $count => $result) {
+                $res = $this->serializer->encodeMessage($result);
+                yield $res;
+            }
+        } catch (\Exception $ex) {
+            throw $this->convertToGoogleException($ex);
+        }
+    }
+
+    /**
+     * Convert a ApiCore exception to a Google Exception.
+     *
+     * @param \Exception $ex
+     * @return Exception\ServiceException
+     */
+    private function convertToGoogleException($ex)
```

Comment:
All new functions should include typehinted parameters and return types. In this case, I think we should make it:
```suggestion
    private function convertToGoogleException(ApiException $ex): ServiceException
```

Reply:
Done.